File: driver/wrappers/transform.cpp
#include "transform.h"
#include "selector.h"
#include "group.h"
#include "geometry_group.h"


NativeTransformWrapper::NativeTransformWrapper(optix::Transform transform) {
    this->transform = transform;
    this->set_destroyable_object(this->transform.get());
}

NativeTransformWrapper::~NativeTransformWrapper() {
    if(this->transform.get() != 0) this->transform->destroy();
}

void NativeTransformWrapper::set_matrix(bool transpose, const boost::numpy::ndarray& array) {
    if (array.get_dtype() != boost::numpy::dtype::get_builtin<float>()) {
        PyErr_SetString(PyExc_ValueError, "Transformation matrix must have float dtype");
        boost::python::throw_error_already_set();
    }

    if (array.get_nd() != 2) {
        PyErr_SetString(PyExc_ValueError, "Transformation matrix must be two dimensional");
        boost::python::throw_error_already_set();
    }

    if (array.shape(0) != 4 || array.shape(1) != 4) {
        PyErr_SetString(PyExc_ValueError, "Transformation matrix must be 4 by 4");
        boost::python::throw_error_already_set();
    }

    this->transform->setMatrix(transpose, reinterpret_cast<float*>(array.get_data()), NULL);

}

boost::numpy::ndarray NativeTransformWrapper::get_matrix(bool transpose) {
    boost::numpy::dtype dt = boost::numpy::dtype::get_builtin<float>();
    boost::python::tuple shape = boost::python::make_tuple(4,4);
    boost::numpy::ndarray arr = boost::numpy::zeros(shape, dt);
    this->transform->getMatrix(transpose, reinterpret_cast<float*>(arr.get_data()), NULL);
    return arr;
}

void NativeTransformWrapper::set_child_geometry_group(unsigned int index, NativeGeometryGroupWrapper* child) {
    this->transform->setChild(child->get_native());
}

void NativeTransformWrapper::set_child_group(unsigned int index, NativeGroupWrapper* child) {
    this->transform->setChild(child->get_native());
}

void NativeTransformWrapper::set_child_selector(unsigned int index, NativeSelectorWrapper* child) {
    this->transform->setChild(child->get_native());
}

void NativeTransformWrapper::set_child_transform(unsigned int index, NativeGeometryGroupWrapper* child) {
    this->transform->setChild(child->get_native());
}

optix::Transform NativeTransformWrapper::get_native() {
    return this->transform;
}

void NativeTransformWrapper::export_for_python() {
    boost::python::class_<NativeTransformWrapper, boost::python::bases<NativeDestroyableWrapper> >(
                "NativeTransformWrapper",
                "NativeTransformWrapper docstring",
                boost::python::init<optix::Transform>())

            .def("set_matrix", &NativeTransformWrapper::set_matrix)
            .def("get_matrix", &NativeTransformWrapper::get_matrix)
            .def("_set_child_geometry_group", &NativeTransformWrapper::set_child_geometry_group)
            .def("_set_child_group", &NativeTransformWrapper::set_child_group)
            .def("_set_child_selector", &NativeTransformWrapper::set_child_selector)
            .def("_set_child_transform", &NativeTransformWrapper::set_child_transform);
}

File: pyoptix/objects/transform.py
import numpy
from pyoptix._driver import NativeTransformWrapper
from pyoptix.objects.shared.optix_object import OptixObject
from pyoptix.objects.shared.optix_parent import OptixParent


class TransformObj(NativeTransformWrapper, OptixObject, OptixParent):

    def __init__(self, native, context):
        OptixObject.__init__(self, context, native)
        NativeTransformWrapper.__init__(self, native)

        from pyoptix.objects.geometry_group import GeometryGroupObj
        from pyoptix.objects.group import GroupObj
        from pyoptix.objects.selector import SelectorObj
        allowed_children = [GeometryGroupObj, GroupObj, SelectorObj, TransformObj]

        OptixParent.__init__(self, allowed_children)

        self.transpose = False

    def set(self, matrix, column_major=False):
        self.transpose = column_major

        if isinstance(matrix, numpy.ndarray) and isinstance(matrix.dtype, numpy.float32):
            self.set_matrix(self.transpose, matrix)
        else:
            self.set_matrix(self.transpose, numpy.matrix(matrix, dtype=numpy.float32))

    def get(self, column_major=None):
        transpose = column_major if column_major is not None else self.transpose
        return self.get_matrix(transpose)

    def get_child_count(self):
        return 1

    def _set_child_count(self, count):
        pass

    def set_matrix(self, transpose, matrix):
        matrix = numpy.array(matrix, dtype=numpy.float32)

        if matrix.shape != (4, 4):
            raise ValueError('Transformation matrix must be 4 by 4')

        self._set_matrix(transpose, matrix.tolist())

    def get_matrix(self, transpose):
        return numpy.array(self._get_matrix(transpose), dtype=numpy.float32)
